### PR TITLE
P3168R1: use a non-pointer iterator type

### PR DESCRIPTION
**Pull request by neatudarius**
P3168R1: use a non-pointer iterator type . Follow-up PR after #9 for suggestion made by @brevzin (issue #14) .

Updates:
* `include/Beman/Optional26/detail/iterator.hpp`: Adopt [normal_iterator](https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/stl_iterator.h#L1030-L1338) implementation from GCC and create a simplified version into `beman::optional::detail:normal_iterator`.
* `include/Beman/Optional26/optional.hpp`: Replace implementation for `optional<T/T&>::iterator/const_iterator`: `T*/const*` -> `detail::normal_iterator<T/const T*, optional>` .
* Add more tests related to iterators.

~~Notes:~~
* ~~This should be a very small PR (interesting diff in `optional.hpp`, also possible in `iterator.hpp`).~~
* ~~A long-term implementation for the iterator type will be done in #17. Now we just want something good enough to be able to make experiments for LWG@Saint Louis.~~
L.E.
As discussed in [this thread](https://github.com/beman-project/Optional26/pull/18#discussion_r1647675872), I copied the stl_interfaces Boost implementation into our tree (#17).

### CODE REVIEW

File: src/Beman/Optional26/tests/__utils.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#ifndef BEMAN_OPTIONAL26_OPTIONAL_TESTS__UTILS_HPP
```
Comment (steve-downey) [inline src/Beman/Optional26/tests/__utils.h:1]: utils is a sure to collide name for things, and the names of types in here are, too. This should be beman/optional/test/utils.hpp or something similar? 
Namespace shouldn't be just ::test, or when this is composed into a larger group it's going to collide when linked, even if no TU ends up with two "::test::no_default_ctor" types.

File: src/Beman/Optional26/tests/optional_range_support.t.cpp [hunk 2]
```diff
@@ -213,7 +180,7 @@ TEST(RangeSupportTest, LoopOverEmptyRange) {
     beman::optional::optional<int> empty;
     ASSERT_FALSE(empty.has_value());
 
-    for (auto _ : empty) {
+    for ([[maybe_unused]] auto _ : empty) {
```
Comment (steve-downey) [inline src/Beman/Optional26/tests/optional_range_support.t.cpp:183]: Does this make the compiler less unhappy about _ not being used?

File: include/Beman/Optional26/detail/__iterator.hpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
+// include/Beman/Optional26/detail/__iterator.hpp                                            -*-C++-*-
+#ifndef BEMAN_OPTIONAL26_OPTIONAL_DETAIL___ITERATOR_HPP
+#define BEMAN_OPTIONAL26_OPTIONAL_DETAIL___ITERATOR_HPP
+
+#include <iterator>
+#include <type_traits>
+
+namespace beman::optional {
```
[the remark this answers is not in the document]
Comment (neatudarius) [reply include/Beman/Optional26/detail/__iterator.hpp:8]: Will change to `detail/iterator.cpp`.

File: src/Beman/Optional26/tests/detail/__iterator.t.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
+#include <Beman/Optional26/detail/__iterator.hpp>
+
+/**
+ * This file contains tests for normal_iterator.
+ */
+#include <gtest/gtest.h>
+#include "../__utils.h"
```
[the remark this answers is not in the document]
Comment (neatudarius) [reply src/Beman/Optional26/tests/detail/__iterator.t.cpp:7]: done

File: src/Beman/Optional26/tests/optional_range_support.t.cpp (hunk 2 above)
Comment (neatudarius) [reply src/Beman/Optional26/tests/optional_range_support.t.cpp:187]: At least for some compilers :D

File: include/Beman/Optional26/detail/iterator.hpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
+// include/Beman/Optional26/detail/iterator.hpp                                            -*-C++-*-
+#ifndef BEMAN_OPTIONAL26_OPTIONAL_DETAIL_ITERATOR_HPP
+#define BEMAN_OPTIONAL26_OPTIONAL_DETAIL_ITERATOR_HPP
+
+#include <iterator>
+#include <type_traits>
+
+namespace beman::optional::detail {
+// Implementation inspired by libstdc++'s normal_iterator
+// (https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/stl_iterator.h)
```
Comment (camio) [inline include/Beman/Optional26/detail/iterator.hpp:10]: Some of this code looks exactly like libstdc++'s class. libstdc++'s copyright and license should be reflected in this file.

File: src/Beman/Optional26/tests/detail/iterator.t.cpp [hunk 6]
```diff
@@ -0,0 +1,178 @@
+/**
```
Comment (camio) [inline src/Beman/Optional26/tests/detail/iterator.t.cpp:1]: This file needs license information per [FILE.LICENSE_ID].

File: include/Beman/Optional26/detail/iterator.hpp (hunk 5 above)
Comment (camio) [reply include/Beman/Optional26/detail/iterator.hpp:10]: Hrm, and libstdc++'s license is not a Beman approved license.

Comment (JeffGarland) [reply include/Beman/Optional26/detail/iterator.hpp:10]: I would suggest using boost.iterator or stl_interfaces (this needs c++20 I believe. https://github.com/boostorg/stl_interfaces

Note that stl_interfaces is proposed for c++26 as well so this would be a good use case for it.

File: src/Beman/Optional26/tests/detail/iterator.t.cpp [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+ * This file contains tests for beman::optional::detail::normal_iterator.
+ */
```
Comment (JeffGarland) [inline src/Beman/Optional26/tests/detail/iterator.t.cpp:3]: Why do we have a tests/detail here -- shouldn't it just be tests?  I don't think we need to 'hide' some sort of implementation details for test code.

Comment (steve-downey) [reply src/Beman/Optional26/tests/detail/iterator.t.cpp:3]: If we're not putting the tests in the directory with the code, putting them in a mechanically derivable location makes it easier / possible to teach an IDE or editor to open related files. 
This is a lot of ceremony that seems unneeded on a project with one real file, but with larger projects knowing where things are and go makes it much easier to navigate.

Comment (steve-downey) [reply src/Beman/Optional26/tests/detail/iterator.t.cpp:3]: The library is currently building with std 23. I think it would probably work with 20. Deducing this is the "killer feature" for  libraries in 23, but it's not used here. 

Writing in 20 style makes translation to and from standard wording *very* straightforward.

Comment (JeffGarland) [reply src/Beman/Optional26/tests/detail/iterator.t.cpp:3]: I'm saying this test file (and all) should be in Optional26/tests -- NOT Optional26/tests/detail. I'm not following why we need that extra directory level and then basically leaving tests empty?

Comment (neatudarius) [reply src/Beman/Optional26/tests/detail/iterator.t.cpp:3]: I prefer to have a tree identical for production code and tests code.

File: include/Beman/Optional26/detail/iterator.hpp (hunk 5 above)
Comment (neatudarius) [reply include/Beman/Optional26/detail/iterator.hpp:10]: @JeffGarland , pleaze check issue linked in the PR description

Comment (neatudarius) [reply include/Beman/Optional26/detail/iterator.hpp:10]: Ok. I offered to help @tzlaine to important his iterator to Beman. I understood the wording may change.

Comment (JeffGarland) [reply include/Beman/Optional26/detail/iterator.hpp:10]: Well I would hope the wording would be the same, but don't stop!

File: src/Beman/Optional26/tests/detail/iterator.t.cpp (hunk 7 above)
Comment (JeffGarland) [reply src/Beman/Optional26/tests/detail/iterator.t.cpp:3]: Ok well I consider it strange. What I was missing is that already on main is the public interface tests.

Comment (neatudarius) [reply src/Beman/Optional26/tests/detail/iterator.t.cpp:3]: yes, I merged it in a previous PR with initial implementation using T*

Comment (neatudarius) [reply src/Beman/Optional26/tests/detail/iterator.t.cpp:3]: I propose to move this discussion into a Beman Standard thread. To be honest, I'm happy to change it to any convention if we have a guideline.

File: include/Beman/Optional26/detail/iterator.hpp (hunk 5 above)
Comment (neatudarius) [reply include/Beman/Optional26/detail/iterator.hpp:10]: @camio , @JeffGarland , I made an experiment directly using `boost/stl_interfaces` as an external dep to `Beman.Optional26`. I was able to build and test with clang-18 (not sure yet why gcc13/14 won't work). 

Diff can be seen in https://github.com/neatudarius/Beman.Optional26/pull/2/.

Is this direction OK for everyody?

Comment (camio) [reply include/Beman/Optional26/detail/iterator.hpp:10]: Hrm, it'd be great to not pull in a huge Boost dependency. Is it feasible to put a copy of that code in the detail namespace instead?

Comment (neatudarius) [reply include/Beman/Optional26/detail/iterator.hpp]: I copied 3 files from that Boost library in a6c6fa3bae9367733ebf354f4b7ae24c7da6965b

```shell
tree include/Beman/Optional26/detail 
include/Beman/Optional26/detail
├── iterator.hpp
└── stl_interfaces
    ├── config.hpp
    ├── fwd.hpp
    └── iterator_interface.hpp
 ```
 
 Next commit (f4eaa880ea6397907dc3636f215c69dc3b2c7189) renames the namespace / macros to be specific to `beman::optional::detail`.
 
 @camio , can you take a look?

File: src/Beman/Optional26/tests/detail/iterator.t.cpp (hunk 6 above)
Comment (neatudarius) [reply src/Beman/Optional26/tests/detail/iterator.t.cpp:4]: Please check - updated entire tree in 416ec70bd1e9da43702b29f4c01a231d22c2069d

File: include/Beman/Optional26/detail/iterator.hpp (hunk 5 above)
Comment (JeffGarland) [reply include/Beman/Optional26/detail/iterator.hpp]: The better convention here is probably to put the stl_interfaces under external in the repository, but this is probably fine as well.  Ideally any macros / namespaces needed in stl interfaces would go back into stl_interfaces primary repo so that in the end a fetch content cmake would be able to grab the dependency instead of copying here.  I took a brief look at the iterator and it looked fine, but will try to dig deeper later.

So you're spearheading a number of technical things for Beman here -- I expect to spend some time on Thursday night discussing this project at the Beman night session.  Good stuff :)

Comment (neatudarius) [reply include/Beman/Optional26/detail/iterator.hpp]: We discussed 2 approaches in this thread:

1. I've initially attached a commit from other branch were I had `external/stl_interfaces` - check 7457c76f9254b8ea494c3d6e790865c15f54b2d2. In this approach, I tried not change at all the Boost code (used `extern/stl_interfaces/include` as include directory in `src/Beman/Optional26/CMakeLists.txt`, and was able to directly include headers from `boost/stl_interfaces` in `iterator.hpp` in this repo).  

2. @camio  suggested that we better copy 3 headers from `boost/stl_interfaces` in `include/Beman/Optional26/detail` and then we rename whatever means boost names (namespaces, macros etc) to be sure that we don't have any interaction with other boost library versions installed on target system - basically, make `Beman.Optionak26` a standalone library, no production deps. When we'll have a `Beman.STLInterfaces` library available we can remove copied files.

> The better convention here is probably to put the stl_interfaces under external in the repository, but this is probably fine as well. Ideally any macros / namespaces needed in stl interfaces would go back into stl_interfaces primary repo so that in the end a fetch content cmake would be able to grab the dependency instead of copying here. 

@JeffGarland , I have a clarifying question here: are you aiming for option1 or option 2 or a mixt of them/new option 3? The `fetch content` suggestion seems to be `option 2` with no extra changes :D

Comment (JeffGarland) [reply include/Beman/Optional26/detail/iterator.hpp]: For now, I'm basically fine with either -- but I think our long term target should be fetch content with Beman stuff incorporated into the Boost source repo. Unless it is @tzlaine vision to port stl_interfaces to Beman specific repo -- but I don't see that as being needed. I very much prefer the external location (copied or fetched) since that makes it clear that this is work from elsewhere that is technically incidental to the implementation here. I certainly expect this won't be our last Boost dependency -- still many useful tools over there for library building.

Comment (neatudarius) [reply include/Beman/Optional26/detail/iterator.hpp]: > So you're spearheading a number of technical things for Beman here -- I expect to spend some time on Thursday night discussing this project at the Beman night session. Good stuff :)

Do we have something on agenda? :D please share a link.

Comment (JeffGarland) [reply include/Beman/Optional26/detail/iterator.hpp]: https://wiki.edg.com/bin/view/Wg21stlouis2024/EveningSessions

Are you going to be there?

Comment (neatudarius) [reply include/Beman/Optional26/detail/iterator.hpp]: ok. that's 3:30 AM Bucharest time :( I'll try...

File: include/Beman/Optional26/detail/stl_interfaces/iterator_interface.hpp [hunk 8]
```diff
@@ -0,0 +1,986 @@
+// include/Beman/Optional26/detail/stl_interfaces/iterator_interface.hpp -*-C++-*-
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```
Comment (camio) [inline include/Beman/Optional26/detail/stl_interfaces/iterator_interface.hpp:2]: This should be the Boost license identifier. Also, the top-level README should mention that some code is licensed under Boost and will be marked as such.

Comment (camio) [reply include/Beman/Optional26/detail/stl_interfaces/iterator_interface.hpp:2]: Well, unless you could get @tzlaine to give you an "Apache-2.0 WITH LLVM-exception" license to his code. That would make it simpler.

**Comment by steve-downey** [comment]
I had a little time this weekend:
https://github.com/steve-downey/iterator-interface
It's an implementation from the wording and exposition from P2727R4 - Only
"creative" bit was writing the iterator tag calculation.

Did turn up some wording nits, so not totally useless.

File: include/Beman/Optional26/detail/stl_interfaces/iterator_interface.hpp (hunk 8 above)
Comment (tzlaine) [reply include/Beman/Optional26/detail/stl_interfaces/iterator_interface.hpp:2]: This is my official permission to use the Boost.STLInterfaces code under the Apache-2.0 WITH LLVM-exception license.

Comment (neatudarius) [reply include/Beman/Optional26/detail/stl_interfaces/iterator_interface.hpp:2]: That's great, @tzlaine! One more step closer to have a demo for Beman sync in Saint Louis.

Also, please check #21 (local demo with Godbolt and my fork of Beman.Optional26).

**Comment by neatudarius** [comment]
> I had a little time this weekend: https://github.com/steve-downey/iterator-interface It's an implementation from the wording and exposition from P2727R4 - Only "creative" bit was writing the iterator tag calculation. Did turn up some wording nits, so not totally useless.

Nice! Thanks Steve! I think we can start a chat with @tzlaine and see that's the plan for stl_interfaces inside Beman.

File: include/Beman/Optional26/detail/iterator.hpp (hunk 5 above)
Comment (neatudarius) [reply include/Beman/Optional26/detail/iterator.hpp]: Closed this thread as now we have a copy of boost/stl_interfaces sources + permission to use them under "Apache-2.0 WITH LLVM-exception"  licence (https://github.com/beman-project/Optional26/pull/18#discussion_r1651196265).

File: include/Beman/Optional26/detail/iterator.hpp [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+//
+// @tparam T - The type of the elements the iterator points to.
+// @tparam Container - The type of the container the iterator points to. This parameter exists solely so that different
+// containers using this template can instantiate different types, even if the T parameter is the same.
+template <class T, class Container>
+struct contiguous_iterator : public base_contiguous_iterator<T, Container> {
+    // Alias for the base class.
+    using base_type = base_contiguous_iterator<T, Container>;
+    // Alias for types from the base class.
+    using typename base_type::difference_type;
```
Comment (tzlaine) [inline include/Beman/Optional26/detail/iterator.hpp:41]: It looks like you need difference_type and pointer, but I don't think you need the rest.

**Review by tzlaine — APPROVED** [review]
I added a small note, but it's entirely style.  The rest looks good!

File: include/Beman/Optional26/detail/iterator.hpp (hunk 9 above)
Comment (neatudarius) [reply include/Beman/Optional26/detail/iterator.hpp:41]: right, also need reference. removed unused aliases

File: include/Beman/Optional26/detail/iterator.hpp [hunk 10]
```diff
@@ -59,6 +63,11 @@ struct contiguous_iterator : public base_contiguous_iterator<T, Container> {
     T* m_current;
 };
 
+// Check concepts for the contiguous_iterator class.
```
Comment (steve-downey) [inline include/Beman/Optional26/detail/iterator.hpp:63]: Move to a test. It's small in the scheme of things, but we don't need to check this every time this header is used.

**Review by steve-downey — CHANGES_REQUESTED** [review]

File: include/Beman/Optional26/detail/iterator.hpp (hunk 10 above)
Comment (neatudarius) [reply include/Beman/Optional26/detail/iterator.hpp:63]: We already have extensive tests for these in `src/Beman/Optional26/tests/detail/iterator.t.cpp`:

```shell
    // non-const iterators
    test(beman::optional::detail::contiguous_iterator<int, container>{});
    test(beman::optional::detail::contiguous_iterator<empty, container>{});
    test(beman::optional::detail::contiguous_iterator<no_default_ctor, container>{});
    test(beman::optional::detail::contiguous_iterator<base, container>{});
    test(beman::optional::detail::contiguous_iterator<derived, container>{});

    // const iterators
    test(beman::optional::detail::contiguous_iterator<const int, container>{});
    test(beman::optional::detail::contiguous_iterator<const empty, container>{});
    test(beman::optional::detail::contiguous_iterator<const no_default_ctor, container>{});
    test(beman::optional::detail::contiguous_iterator<const base, container>{});
    test(beman::optional::detail::contiguous_iterator<const derived, container>{});
 ```

I removed this check from the header.

File: include/Beman/Optional26/optional.hpp [hunk 11]
```diff
@@ -321,6 +322,9 @@ class optional {
     using iterator       = detail::contiguous_iterator<T, optional>;       // see [optional.iterators]
     using const_iterator = detail::contiguous_iterator<const T, optional>; // see [optional.iterators]
 
+    static_assert(std::contiguous_iterator<iterator>);
```
[the remark this answers is not in the document]
Comment (neatudarius) [reply include/Beman/Optional26/optional.hpp:325]: also here: we already have tests in `Beman/Beman.Optional26/src/Beman/Optional26/tests/optional_range_support.t.cpp`:

```shell
TEST(RangeSupportTest, IteratorConcepts) {
    const auto test = [](auto&& opt) {
        // The iterator type is the same as the iterator of the optional.
        // e.g. iterator = optional<T>::iterator if opt is optional<T>
        using iterator = typename std::remove_reference_t<decltype(opt)>::iterator;
        // e.g. const_iterator = optional<T>::const_iterator if opt is optional<T>
        using const_iterator = typename std::remove_reference_t<decltype(opt)>::const_iterator;

        ...
        static_assert(std::contiguous_iterator<iterator>);

        ...
        static_assert(std::contiguous_iterator<const_iterator>);
    };

    test(beman::optional::optional<int>{});
    test(beman::optional::optional<int*>{});
    test(beman::optional::optional<empty>{});
    test(beman::optional::optional<no_default_ctor>{});
    test(beman::optional::optional<base>{});
    test(beman::optional::optional<derived>{});
}
```

**Review by steve-downey — APPROVED** [review]

**Review by JeffGarland — APPROVED** [review]

**Review by camio — APPROVED** [review]

**Comment by neatudarius** [comment]
Merged the PR to be able to deploy into Compiler Explorer (#21 ).

Follow-up tweaks will be done in other PRs.
